pkg/helper/webconsole.go

```diff
@@ -0,0 +1,120 @@
+package helper
+
+import (
+	"context"
+	"fmt"
+	"github.com/RHsyseng/operator-utils/pkg/logs"
+	"github.com/RHsyseng/operator-utils/pkg/resource/read"
+	"github.com/RHsyseng/operator-utils/pkg/utils/kubernetes"
+	consolev1 "github.com/openshift/api/console/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strings"
+	"time"
+)
+
+var logger = logs.GetLogger("openshift-webconsole")
+
+// ConsoleLinkText is the text of the consoleLink shown on the webconsole
+const ConsoleLinkText = "APIManager - 3scale"
+// ConsoleLinkNamePrefix is the prefix applied to Console link for system master
+const ConsoleLinkNamePrefix = "system-master-link"
+
+// ConsoleLinkSupported checks if a ConsoleLink CRD exists
+func ConsoleLinkSupported() error {
+	gvk := schema.GroupVersionKind{Group: "console.openshift.io", Version: "v1", Kind: "ConsoleLink"}
+	return kubernetes.CustomResourceDefinitionExists(gvk)
+}
+
+// ReconcileConsoleLink creates a ConsoleLink object if it doesn't already exist
+func ReconcileConsoleLink(ctx context.Context, c client.Client, req *reconcile.Request) {
+	route := getRoute(ctx, c, req)
+	if route == nil {
+		removeConsoleLink(ctx, c, req)
+	} else {
+		consoleLinkName := fmt.Sprintf("%s-%s", ConsoleLinkNamePrefix, req.Namespace)
+		consoleLink := &consolev1.ConsoleLink{}
+		err := c.Get(ctx, types.NamespacedName{Name: consoleLinkName}, consoleLink)
+		if err != nil && apierrors.IsNotFound(err) {
```

Comment:
What happens when error is nil and apierrors is not the IsNotFound error?
It seems we ignore the error in that case. I think when that happens we should log the error and perform a requeue (with a delay if you prefer)

Reply:
- Then if the link exists, it will be updated.
- added logging message.  `else` part

Reply:
For what I see in the current implementation if there's an error we do not do anything. So if there's some temporary error when doing a create/update/delete the action would be simply lost.
I think we should retry in those cases instead of just logging because otherwise the ConsoleLink would never be created/reconciled (unless there's an external event on the route that triggers the reconciliation loop again)

Reply:
if the error situation is not temporary, and it persists, will that cause nonstop requeue? Do we need additional logic prevent this from happening?

Reply:
It will cause a nonstop requeue indeed. A delay requeue can be added if desired(with RequeueAfter), and in any case when lots of events are sent Kubernetes performs aggregation of them and also applies an exponential backoff retry time when it detects lots of them being sent in a rapid succession

Reply:
I made a manual test which makes the deletion will never succeed. I dont see any backoff. I need to kill the app and operator to stop the events. I think this should never happen in normal situation, just for your info..

Reply:
The events would never stop in that case. They should be incrementally taking longer to be received due to the exponential backoff when lots of them are received in a quick succession. Maybe you are not reproducing the exponential backoff because you've added a delay in the requeue that might be big enough to not be considered "event spamming" by the framework :thinking: